File: apps/backend-gateway/src/application/schemas/stadium_schemas.py
"""
StadiumOS AI — Stadium & Sector Pydantic Schemas.

Validation and serialization DTO schemas for Stadium and Sector API endpoints.
"""

from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class SectorCreate(BaseModel):
    """Payload to create a new stadium sector."""

    name: str = Field(..., min_length=2, max_length=100, description="Sector name (e.g. 'North Stand').")
    max_capacity: int = Field(..., gt=0, description="Max safe capacity.")
    warning_threshold_percent: float = Field(default=80.0, ge=50.0, le=100.0)
    critical_threshold_percent: float = Field(default=95.0, ge=70.0, le=100.0)
    centroid_latitude: float = Field(..., ge=-90.0, le=90.0)
    centroid_longitude: float = Field(..., ge=-180.0, le=180.0)
    is_accessible: bool = Field(default=False)

    @field_validator("critical_threshold_percent")
    @classmethod
    def validate_thresholds(cls, val: float, info: object) -> float:
        warning = info.data.get("warning_threshold_percent", 80.0)  # type: ignore[union-attr]
        if val <= warning:
            raise ValueError("Critical threshold must exceed warning threshold.")
        return val


class SectorResponse(BaseModel):
    """Serialized stadium sector data."""

    id: UUID
    stadium_id: UUID
    name: str
    max_capacity: int
    warning_threshold_percent: float
    critical_threshold_percent: float
    centroid_latitude: float
    centroid_longitude: float
    is_accessible: bool

    class Config:
        from_attributes = True


class StadiumCreate(BaseModel):
    """Payload to create a new stadium venue."""

    name: str = Field(..., min_length=2, max_length=150)
    city: str = Field(..., min_length=2, max_length=100)
    country: str = Field(..., min_length=2, max_length=100)
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    total_capacity: int = Field(..., gt=0)
    timezone: str = Field(default="UTC", min_length=2, max_length=50)


class StadiumResponse(BaseModel):
    """Serialized stadium data."""

    id: UUID
    name: str
    city: str
    country: str
    latitude: float
    longitude: float
    total_capacity: int
    timezone: str

    class Config:
        from_attributes = True

File: apps/backend-gateway/src/application/schemas/stadiumSchemas.ts
/**
 * StadiumOS AI — Stadium & Sector API Validation Schemas.
 * 
 * Defines Zod validation schemas for stadium and sector creation payloads.
 */

import { z } from "zod";

export const StadiumCreateSchema = z.object({
  name: z.string().min(2).max(150),
  city: z.string().min(2).max(100),
  country: z.string().min(2).max(100),
  latitude: z.number().min(-90.0).max(90.0),
  longitude: z.number().min(-180.0).max(180.0),
  total_capacity: z.number().int().positive(),
  timezone: z.string().min(2).max(50).default("UTC")
});

export type StadiumCreate = z.infer<typeof StadiumCreateSchema>;

export const SectorCreateSchema = z.object({
  name: z.string().min(2).max(100),
  max_capacity: z.number().int().positive(),
  warning_threshold_percent: z.number().min(50.0).max(100.0).default(80.0),
  critical_threshold_percent: z.number().min(70.0).max(100.0).default(95.0),
  centroid_latitude: z.number().min(-90.0).max(90.0),
  centroid_longitude: z.number().min(-180.0).max(180.0),
  is_accessible: z.boolean().default(false)
}).refine((data) => data.critical_threshold_percent > data.warning_threshold_percent, {
  message: "critical_threshold_percent must exceed warning_threshold_percent",
  path: ["critical_threshold_percent"]
});

export type SectorCreate = z.infer<typeof SectorCreateSchema>;
